File: backend/src/modules/workflow/workflowManager.js
export const WORKFLOW_STEP_CODES = {
  ENTER_SHOWER: 'ENTER_SHOWER',
  HAND_SANITIZATION: 'HAND_SANITIZATION',
  BOOT_SANITIZATION: 'BOOT_SANITIZATION',
  ENTER_CLEAN_ZONE: 'ENTER_CLEAN_ZONE',
}

export const STEP_ZONE_CODES = {
  [WORKFLOW_STEP_CODES.ENTER_SHOWER]: 'shower_room',
  [WORKFLOW_STEP_CODES.HAND_SANITIZATION]: 'handwash_zone',
  [WORKFLOW_STEP_CODES.BOOT_SANITIZATION]: 'boot_disinfection_tray',
}

export const CLEAN_PRODUCTION_ZONES = new Set([
  'gestation_barn',
  'farrowing_barn',
  'weaning_barn',
  'nursery_barn',
  'boar_barn',
  'quarantine_barn',
])

export const ENTRY_CLEAN_ZONE_WORKFLOW = {
  id: 'entry_clean_zone',
  name: 'Quy trình vào vùng sạch',
  steps: [
    WORKFLOW_STEP_CODES.ENTER_SHOWER,
    WORKFLOW_STEP_CODES.HAND_SANITIZATION,
    WORKFLOW_STEP_CODES.BOOT_SANITIZATION,
    WORKFLOW_STEP_CODES.ENTER_CLEAN_ZONE,
  ],
}

export const DEFAULT_WORKFLOW_DEFINITIONS = [ENTRY_CLEAN_ZONE_WORKFLOW]

/** @typedef {{ id: string, name: string, steps: string[], stepZones?: Record<string, string> }} WorkflowDefinition */

/** @param {WorkflowDefinition} definition @param {string} stepCode */
export function zoneForStep(definition, stepCode) {
  if (definition.stepZones?.[stepCode]) {
    return definition.stepZones[stepCode]
  }
  if (stepCode === WORKFLOW_STEP_CODES.ENTER_CLEAN_ZONE) {
    return null
  }
  return STEP_ZONE_CODES[stepCode] || null
}

/** @param {WorkflowDefinition} definition @param {string} zoneCode */
export function stepForZone(definition, zoneCode) {
  for (const stepCode of definition.steps) {
    const mapped = zoneForStep(definition, stepCode)
    if (mapped && mapped === zoneCode) {
      return stepCode
    }
  }
  if (
    CLEAN_PRODUCTION_ZONES.has(zoneCode) &&
    definition.steps.includes(WORKFLOW_STEP_CODES.ENTER_CLEAN_ZONE)
  ) {
    return WORKFLOW_STEP_CODES.ENTER_CLEAN_ZONE
  }
  return null
}

export class WorkflowManager {
  constructor() {
    /** @type {Map<string, WorkflowDefinition>} */
    this._definitions = new Map()
  }

  /** @param {WorkflowDefinition} definition */
  registerDefinition(definition) {
    this._definitions.set(definition.id, definition)
  }

  /** @param {string} workflowId */
  getDefinition(workflowId) {
    return this._definitions.get(workflowId) || null
  }

  listDefinitions() {
    return Array.from(this._definitions.values())
  }

  loadDefaults() {
    for (const definition of DEFAULT_WORKFLOW_DEFINITIONS) {
      this.registerDefinition(definition)
    }
  }
}

let _manager = null

export function initWorkflowManager() {
  _manager = new WorkflowManager()
  _manager.loadDefaults()
  return _manager
}

export function getWorkflowManager() {
  if (!_manager) {
    return initWorkflowManager()
  }
  return _manager
}

File: backend/app/biosecurity_workflow/workflow_manager.py
from __future__ import annotations

import logging

from app.biosecurity_workflow.definitions import DEFAULT_WORKFLOW_DEFINITIONS, WorkflowDefinition

logger = logging.getLogger(__name__)


class WorkflowManager:
    def __init__(self) -> None:
        self._definitions: dict[str, WorkflowDefinition] = {}

    def register_definition(self, definition: WorkflowDefinition) -> None:
        self._definitions[definition.id] = definition

    def get_definition(self, workflow_id: str) -> WorkflowDefinition | None:
        return self._definitions.get(workflow_id)

    def list_definitions(self) -> list[WorkflowDefinition]:
        return list(self._definitions.values())

    def load_defaults(self) -> None:
        for definition in DEFAULT_WORKFLOW_DEFINITIONS:
            self.register_definition(definition)
        logger.info(
            "Biosecurity Workflow Manager loaded %s definitions: %s",
            len(self._definitions),
            ", ".join(self._definitions.keys()),
        )


_manager: WorkflowManager | None = None


def init_workflow_manager() -> WorkflowManager:
    global _manager
    _manager = WorkflowManager()
    _manager.load_defaults()
    return _manager


def get_workflow_manager() -> WorkflowManager:
    if _manager is None:
        return init_workflow_manager()
    return _manager
